src/front/wgsl/mod.rs

```diff
@@ -1559,1180 +1697,348 @@ impl Parser {
-                Ok(Number::I32(num)) => crate::ConstantInner::Scalar {
-                    value: crate::ScalarValue::Sint(num as i64),
-                    width: 4,
-                },
-                Ok(Number::U32(num)) => crate::ConstantInner::Scalar {
-                    value: crate::ScalarValue::Uint(num as u64),
-                    width: 4,
-                },
-                Ok(Number::F32(num)) => crate::ConstantInner::Scalar {
-                    value: crate::ScalarValue::Float(num as f64),
-                    width: 4,
-                },
-                Ok(Number::AbstractInt(_) | Number::AbstractFloat(_)) => unreachable!(),
-                Err(e) => return Err(Error::BadNumber(first_token_span.1, e)),
-            },
-            (Token::Word(name), name_span) => {
-                // look for an existing constant first
-                for (handle, var) in const_arena.iter() {
-                    match var.name {
-                        Some(ref string) if string == name => {
-                            self.pop_rule_span(lexer);
-                            return Ok(handle);
-                        }
-                        _ => {}
-                    }
-                }
-                let composite_ty = self.parse_type_decl_name(
-                    lexer,
-                    name,
-                    name_span,
-                    None,
-                    TypeAttributes::default(),
-                    type_arena,
-                    const_arena,
-                )?;
+        let expr = match name {
+            // bitcast looks like a function call, but it's an operator and must be handled differently.
+            "bitcast" => {
+                lexer.expect_generic_paren('<')?;
+                let start = lexer.start_byte_offset();
+                let to = self.parse_type_decl(lexer, ctx.reborrow())?;
+                let span = lexer.span_from(start);
+                lexer.expect_generic_paren('>')?;
 
                 lexer.open_arguments()?;
-                //Note: this expects at least one argument
-                let mut components = Vec::new();
-                while components.is_empty() || lexer.next_argument()? {
-                    let component = self.parse_const_expression(lexer, type_arena, const_arena)?;
-                    components.push(component);
+                let expr = self.parse_general_expression(lexer, ctx.reborrow())?;
+                lexer.close_arguments()?;
+
+                ast::Expression::Bitcast {
+                    expr,
+                    to,
+                    ty_span: span,
                 }
-                crate::ConstantInner::Composite {
-                    ty: composite_ty,
-                    components,
+            }
+            // everything else must be handled later, since they can be hidden by user-defined functions.
+            _ => {
+                let arguments = self.parse_arguments(lexer, ctx.reborrow())?;
```

Comment:
This needs to record a dependency as well (and there should be a test for this):

```suggestion
                ctx.unresolved.insert(ast::Dependency {
                    ident: name,
                    usage: name_span,
                });

                let arguments = self.parse_arguments(lexer, ctx.reborrow())?;
```